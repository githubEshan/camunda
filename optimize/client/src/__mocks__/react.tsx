/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH under
 * one or more contributor license agreements. See the NOTICE file distributed
 * with this work for additional information regarding copyright ownership.
 * Licensed under the Camunda License 1.0. You may not use this file
 * except in compliance with the Camunda License 1.0.
 */

// @ts-expect-error
export * from 'react';
export {default} from 'react';

const outstandingEffects: (() => void)[] = [];
const outstandingCleanups: (() => void)[] = [];

export const useEffect = (fn: () => (() => void) | void) => outstandingEffects.push(fn);
export const runLastEffect = () => {
  if (outstandingEffects.length) {
    const cleanup = outstandingEffects.pop()?.();

    if (cleanup) {
      outstandingCleanups.push(cleanup);
    }
  }
};

export const runLastCleanup = () => {
  if (outstandingCleanups.length) {
    outstandingCleanups.pop()?.();
  }
};

export const runAllEffects = () => {
  const numberOfEffects = outstandingEffects.length;

  for (let i = 0; i < numberOfEffects; i++) {
    const cleanup = outstandingEffects.shift()?.();

    if (cleanup) {
      outstandingCleanups.push(cleanup);
    }
  }
};

beforeEach(() => {
  outstandingEffects.length = 0;
});

export const runAllCleanups = () => {
  const numberOfEffects = outstandingCleanups.length;

  for (let i = 0; i < numberOfEffects; i++) {
    outstandingCleanups.shift()?.();
  }
};
