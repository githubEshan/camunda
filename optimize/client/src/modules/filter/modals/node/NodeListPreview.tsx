/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH under
 * one or more contributor license agreements. See the NOTICE file distributed
 * with this work for additional information regarding copyright ownership.
 * Licensed under the Camunda License 1.0. You may not use this file
 * except in compliance with the Camunda License 1.0.
 */

import {ReactNode} from 'react';
import {Tag} from '@carbon/react';

import {ModdleElement} from 'components';
import {t} from 'translation';

interface NodeListPreviewProps {
  nodes: ModdleElement[];
  operator?: string;
  type: string;
}

export default function NodeListPreview({nodes, operator, type}: NodeListPreviewProps) {
  const previewList: JSX.Element[] = [];
  const createOperator = (name: ReactNode) => {
    return <span className="previewItemOperator"> {name} </span>;
  };

  nodes.forEach((selectedNode, idx) => {
    previewList.push(
      <li key={idx} className="previewItem">
        <span>
          {' '}
          {idx !== 0 &&
            createOperator(
              operator === 'not in'
                ? t('common.filter.list.operators.and')
                : t('common.filter.list.operators.or')
            )}
          <b>{selectedNode.name || selectedNode.id}</b>
        </span>
      </li>
    );
  });

  const parameterName = t(
    'common.filter.nodeModal.preview.' + (operator === 'not in' ? 'notExecutedFlowNodes' : type)
  ).toString();

  return (
    <>
      <Tag type="blue" className="parameterName">
        <span title={parameterName}>{parameterName}</span>
      </Tag>
      <ul className="previewList filterText">{previewList}</ul>
    </>
  );
}
