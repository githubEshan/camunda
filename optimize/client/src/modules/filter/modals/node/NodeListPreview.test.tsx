/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH under
 * one or more contributor license agreements. See the NOTICE file distributed
 * with this work for additional information regarding copyright ownership.
 * Licensed under the Camunda License 1.0. You may not use this file
 * except in compliance with the Camunda License 1.0.
 */

import {shallow} from 'enzyme';
import {Tag} from '@carbon/react';

import NodeListPreview from './NodeListPreview';

const props = {
  nodes: [{id: 'bar', name: 'foo'}],
  operator: 'in',
  type: 'executedFlowNodes',
};

it('should create preview of nodes', () => {
  const node = shallow(<NodeListPreview {...props} />);

  expect(node).toMatchSnapshot();
});

it('should show the id of the flow node if the name is null', () => {
  const node = shallow(<NodeListPreview {...props} nodes={[{id: 'bar', name: undefined}]} />);

  expect(node.find('b')).toMatchSnapshot();
});

it('should create preview of selected nodes linked by or', () => {
  const flowNode1 = {
    name: 'foo',
    id: 'bar',
  };

  const flowNode2 = {
    name: 'foo',
    id: 'bar',
  };
  const node = shallow(<NodeListPreview {...props} nodes={[flowNode1, flowNode2]} />);

  expect(node).toMatchSnapshot();
});

it('should create preview of selected nodes linked by nor', () => {
  const flowNode1 = {
    name: 'foo',
    id: 'bar',
  };

  const flowNode2 = {
    name: 'foo',
    id: 'bar',
  };

  const changedProps = {
    nodes: [flowNode1, flowNode2],
    operator: 'not in',
  };

  const node = shallow(<NodeListPreview {...props} {...changedProps} />);

  expect(node).toMatchSnapshot();
});

it('should show executing node filter', () => {
  const node = shallow(
    <NodeListPreview {...props} operator={undefined} type="executingFlowNodes" />
  );

  expect(node.find(Tag)).toIncludeText('Running');
});
