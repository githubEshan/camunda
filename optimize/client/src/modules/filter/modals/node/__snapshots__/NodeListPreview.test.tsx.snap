// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should create preview of nodes 1`] = `
<Fragment>
  <ForwardRef(Tag)
    className="parameterName"
    type="blue"
  >
    <span
      title="Running, canceled, or completed flow nodes"
    >
      Running, canceled, or completed flow nodes
    </span>
  </ForwardRef(Tag)>
  <ul
    className="previewList filterText"
  >
    <li
      className="previewItem"
      key="0"
    >
      <span>
         
        <b>
          foo
        </b>
      </span>
    </li>
  </ul>
</Fragment>
`;

exports[`should create preview of selected nodes linked by nor 1`] = `
<Fragment>
  <ForwardRef(Tag)
    className="parameterName"
    type="blue"
  >
    <span
      title="Not executed flow nodes"
    >
      Not executed flow nodes
    </span>
  </ForwardRef(Tag)>
  <ul
    className="previewList filterText"
  >
    <li
      className="previewItem"
      key="0"
    >
      <span>
         
        <b>
          foo
        </b>
      </span>
    </li>
    <li
      className="previewItem"
      key="1"
    >
      <span>
         
        <span
          className="previewItemOperator"
        >
           
          and
           
        </span>
        <b>
          foo
        </b>
      </span>
    </li>
  </ul>
</Fragment>
`;

exports[`should create preview of selected nodes linked by or 1`] = `
<Fragment>
  <ForwardRef(Tag)
    className="parameterName"
    type="blue"
  >
    <span
      title="Running, canceled, or completed flow nodes"
    >
      Running, canceled, or completed flow nodes
    </span>
  </ForwardRef(Tag)>
  <ul
    className="previewList filterText"
  >
    <li
      className="previewItem"
      key="0"
    >
      <span>
         
        <b>
          foo
        </b>
      </span>
    </li>
    <li
      className="previewItem"
      key="1"
    >
      <span>
         
        <span
          className="previewItemOperator"
        >
           
          or
           
        </span>
        <b>
          foo
        </b>
      </span>
    </li>
  </ul>
</Fragment>
`;

exports[`should show the id of the flow node if the name is null 1`] = `
<b>
  bar
</b>
`;
